Overhauled

PetrK39 opened the pull request:
Basically I finished refactoring,
only a few key features and a final cleanup are left
- [x] Advanced Blacklist
- [x] Dark theme
    - [ ] Custom light/dark theme
- [ ] Make script more *lazy*
    - [ ] Make infinite scroll lazy
    - [x] Make blacklist lazy
    - [ ] Make every tweak lazy 

~~Later I will add info on the internals and how to add new features~~
Please make sure everything works correctly in all your usecases
(You can now configure them via the Overhaul dialog in the top bar, including without reloading the page.)

---

# Adding new preferences
- Open `gelbooru-overhaul.configManager.js`
- Find `getDefaultConfig()`
It returns the following config object
```js
{string: PreferenceCatgory}
```
Where string is Key
Where PreferenceCategory is 
```js
 /**
  * @typedef  PreferenceCategory 
  * @type     {Object}
  * @property {string}                          name     Displayed name in config window
  * @property {Object.<string, PreferenceItem>} items    Children preference items
  * /
 
{
    name: string,
    items: {string, PreferenceItem}
}
```
Where items.string is Key
Where PreferenceItem is
```js
 /**
  * @typedef  PreferenceItem
  * @type     {Object}
  * @property {number | string | boolean} value          Value of preference
  * @property {string}                    name           Displayed name in config window
  * @property {string}                    description    Description displayed in config window
  * @property {boolean}                   [locked=false] Determines if preference should be available for editing
  * /
 
{
    name: string
    description: string
    locked: boolean
    number | string | boolean: value
}
```
- Add new category if needed
- Add new item

The keys can be any string, but I prefer to keep the category structure
`Category1.Item1`, `Category1.Item2`, `Category2.Item1`

Preference value can be accessed using 
```js
configManager.findValueByKey(key: string)
```

# Adding new features

- Open `gelbooru-overhaul.user.js`
- Find `main()`
- Add new line 
```js
configManager.addUpdateListener(key: string, listener);
```
Where listener is
```js
function(value: number | string | boolean)
```
which will be caused by the change of preference value and on page load
If your setting is only used as a value in another function, there is no need to track its change
The listeners should be arranged so that any of your preference tweak can be turned off and the page can be restored to its original appearance

You also may want to check `currentPageType` from `utils.pageTypes`

If your preference sets css variables, see `applyCssVariableGoCollapseSidebar()` for example on how to do it
If your preference has dependent preferences see `applyTweakEnlargeOnHover()` for example

You can access page's thumbnails with `utils.getThumbnails()` which returns `HTMLImageElement[]` (posts for Gallery page, 'More Like This' for Post page)

Config window automatically display your new preference but there is no validation and fool proof yet
Note that values updates live

For thumbnail tweaks on Gallery page there is also `infiniteScrolling.addUpdateListener(listener)`
Where listener is 
```js
function(value: number | string | boolean, thumbs: HTMLImageElement[])
```
Should be applied only for new thumbs

Script starts on DOM ready so there is no need to check this

---

I don't urge you to take the pool requester right now, but don't forget to replace `@resource css` with your repository
There are many improvements to the original script, something that was impossible before now works fine. You can check out the commit comments

PetrK39 committed:
Refactor Step 1: Configuration Storage
Basic configuration management, storing in the userscript extension storage, configuration versions and migrations

Guide to adding configuration properties
- Modify `ConfigManager.getDefaultConfig()`. Add/rename/delete properties and increment `configVersion`.
- Increment `currentConfigVersion` in `ConfigManager`.
- Add a migration step in `ConfigManager.migrateConfig`. Specify changes, update renamed or dependent properties, and delete properties no longer in use. See commented out example

PetrK39 committed:
Refactor Step 1.5
Some configManager fixes
The migration example now more clear

Debug log feature

Page type detection in order not to make unnecessary changes

PetrK39 committed:
Refactor Step 2: Reimplementation
Fully reimplemented collapsible sidebar function
Added customisation of the sidebar appearance
Resizing the window now works fine
The content of the sidebar is independent of the window size
All changes are applied non-destructively by css classes, so they can be undone even without reloading the page

PetrK39 committed:
made the debugLog a bit nicer

PetrK39 committed:
Test external css file

PetrK39 committed:
Revert "Test external css file"

This reverts commit 97e5c7b5f480d5596110f8082afd05fb1c7597a4.
Not possible to use css with string literals without using unsafe 'eval' and 'new Function'

PetrK39 committed:
Refactor Step 2: Reimplementation
Implemented post fit features
- center
- fit horizontally
- fit vertically on expand
- fit vertically on narrow window
- autoscroll to fit post

PetrK39 committed:
Fixed auto scroll for not cached images
But it can be annoying

PetrK39 committed:
Refactor Step 2: Reimplementation
Reimplemented thumbnail enlarge
Implemented preventing scaled images from going off screen
Added tiny corner rounding for thumbnails

PetrK39 committed:
Refactor Step 2: Reimplementation
All files now have highres thumbnails
Even video. Unfortunately, there is no easy way to make static previews for gif and apng
With RMB any post can be downloaded: video/gif/picture from the gallery/more like this/post page itself
The quick download supports basic substitution, e.g. %postId% - %artist% [%character%] - %md5%
Optionally, a text file can also be created that contains tags
post data is stored in a limited cache.
since there is no particular benefit to storing the cache and potentially performance issues when the cache size = maximum size,
when the maximum is reached it is simply discarded
added the ability to remove the title of the gallery images, it seems, its presence leads to a decrease in performance and visual jerkiness

PetrK39 committed:
started to implement the config window
rough template for config window
some code format

PetrK39 committed:
Fixed performance of EnlargeOnHover
Now thumbnails has high res source only when you hover over it
config window fix
moved all addEventListentr to named functions
adjustments to post fit
expand link now allows to toggle H-V fit
fixed context menu with fastDL

PetrK39 committed:
preventEnlargeOffScreen adjustments

PetrK39 committed:
Big work on config window
Styled and templated
fastDL filePattern now supports custom separator

PetrK39 committed:
Fixed a floating bug
The config window did not detect the background color correctly

PetrK39 committed:
There is no more need for a lodash library

PetrK39 committed:
Image loading indicator
Added rough loader
Thumbs stays in high res until they load
so let the browser cache do its job

PetrK39 committed:
Changed how to apply scale to previews
Now the loading indicator is always aligned

PetrK39 committed:
Another floating DOM bug

PetrK39 committed:
CSS Cleanup
External css possible thanks to css vars
Cleaned duplication and overqualified css
Organized descriptions

PetrK39 committed:
Expand post and toggle fit with click on image
some css tweaks

PetrK39 committed:
Refactor Step 3: Cleanup
A lot of work to bring the script to the way I wanted it to look
Started adding JSDoc, as the script is getting quite complex
Too many changes to explain, hopefully checked everything I could

PetrK39 committed:
sidebar background fix

PetrK39 committed:
InfiniteScroll
and a few related QOL functions like top paginator and go to top
History can work wierd

PetrK39 committed:
Small fix
revert for applyTweakPaginatorOnTop

PetrK39 committed:
InfScroll fix
Fixed the way tweaks are reapplied during infinite scrolling

Enchoseon commented:
Wow, you've fixed so many undocumented bugs that it makes me feel guilty for not bothering to mention any of them anywhere. I can confirm that infinite scroll, collapsible sidebar, thumbnails are working on Tampermonkey 4.17.6161 for Firefox.

Also noticed that the tags in the additional .txt file from FastDL were missing their newlines (e.g. post id `7724955` downloaded a text file with `artist:smile (qd4nsvik)series:originalmeta:highresblue skycablecloudcloudy skyfantasyglowinglandscapelight particlesmoonno humansoceanreflectionreflective watersceneryshadowskystar (sky)starry skysunrisetowertree`), while doing the same on Chrome kept the newlines.

I was able to get the newlines to work in Firefox by changing line [1206](https://github.com/PetrK39/gelbooru-overhaul-userscript/blob/refactoring/gelbooru-overhaul.user.js#L1206) from `elem.href = "data:text," + text;` to `elem.href = "data:text," + encodeURIComponent(text);`—but I'm not sure if this would break other stuff.

In regards to complexity, what do you think about splitting things into separate files and using a bunch of @require's?

Enchoseon commented on gelbooru-overhaul.user.js:
```diff
@@ -9,485 +9,1551 @@
+                postCache = {};
+
+            if (!postCache[postId]) {
+                fetch("https://gelbooru.com/index.php?page=post&s=view&id=" + postId)
```
```suggestion
                fetch("https://" + window.location.host + "/index.php?page=post&s=view&id=" + postId)
```
Add support for ja.gelbooru.com

Enchoseon commented on gelbooru-overhaul.user.js:
```diff
@@ -9,485 +9,1551 @@
+                text = text.replaceAll("\n\n", "\n");
+
+                let elem = document.createElement("a");
+                elem.href = "data:text," + text;
```
```suggestion
                elem.href = "data:text," + encodeURIComponent(text);
```
Fix newlines in the FastDL text file not working in Firefox.

*(Confirmed working on Tampermonkey 4.17.6161 for Chrome and Firefox, but maybe quadruple-check that this screw up any edge-cases)*

Enchoseon commented on gelbooru-overhaul.user.css:
```diff
@@ -0,0 +1,267 @@
+.go-config-window footer input {
+    border: 1px solid;
+}
+
```
```suggestion
.go-config-window footer input:hover {
    background: lightgrey;
}
.go-config-window footer input:active {
    opacity: 0.8;
    transition: 42ms;
}

```
Add a little feedback to the config footer buttons.

PetrK39 replied:
`.go-config-window footer input:hover` causes a rather annoying artifact: the Close button is visible for a few ms after the window is closed
`.go-config-window footer input:active` alone seems to create the same effect but does not result in an artifact
Another option is to add a dissolve effect to the entire div.

PetrK39 replied:
it seems `:hover` is just invisible in dark mode due to gray
Since I'm trying to use the styles that are already on gelbooru, I would like to avoid another calculated color
I'll see what I can do about it later.
The buttons on gelbooru just aren't as beautiful.

PetrK39 replied:
remove fade-in is also a simple solution
```
transition: 142ms;
transition-timing-function: ease-out;
```

PetrK39 replied:
Done https://github.com/Enchoseon/gelbooru-overhaul-userscript/pull/2/commits/6745f5cf2da5f4e69640d36e56e465b59964b57b

Enchoseon commented on gelbooru-overhaul.user.css:
```diff
@@ -0,0 +1,267 @@
+    margin-left: 4px;
+    margin-bottom: 6px;
+    padding: 0 6px;
+    width: -webkit-fill-available;
```
```suggestion
    width: -webkit-fill-available;
    width: -moz-available;
```
Firefox parity.

PetrK39 committed:
Update gelbooru-overhaul.user.js

Suggested change 
Add support for ja.gelbooru.com

Co-authored-by: Encho <38841444+Enchoseon@users.noreply.github.com>

PetrK39 committed:
Update gelbooru-overhaul.user.js

Suggested change 
Fix newlines in the FastDL text file not working in Firefox.

Co-authored-by: Encho <38841444+Enchoseon@users.noreply.github.com>

PetrK39 committed:
Update gelbooru-overhaul.user.css

Suggested change
Firefox parity.

Co-authored-by: Encho <38841444+Enchoseon@users.noreply.github.com>

PetrK39 commented:
There is also a rare bug where when loading the next page with infinite scrolling, the content is added twice
Have you experienced this?
And it seems that the "load" event does not firing on GIF images, causing them to stay animated after "background" loading in the preview

PetrK39 commented:
Breaking up the script into separate files would be a great idea
Especially separate large independent modules like ConfigManager, advanced blacklist and theme manager
But, this would require me to learn more about JavaScript and meanwhile, I want to implement basic functionality

PetrK39 committed:
Make Gelbooru buttons more intractive

PetrK39 committed:
Gallery/More Like This thumbnail size
can be customized using config window

PetrK39 committed:
Changed config window buttons behaviour
Added titles
Close button now cancels unsaved changes
Save button now closes config window

PetrK39 committed:
collapsible sidebar tweaks
to allow wide collapsed sidebars

PetrK39 commented:
<details>
  <summary>Screenshots</summary>
<img src="https://user-images.githubusercontent.com/60184450/192804999-74ce2472-14f8-4968-ad51-d7b10d1fa543.png" width="300">
<img src="https://user-images.githubusercontent.com/60184450/192805010-5a709ac1-f34c-494d-a6fa-4c73e03649b4.png" width="300">
<img src="https://user-images.githubusercontent.com/60184450/192805013-e0933b63-5a16-47af-ae2d-986e035bec1b.png" width="300">
<img src="https://user-images.githubusercontent.com/60184450/192805022-6195213e-47d9-4a28-87d9-c3a1e4c5dcd6.png" width="300">
</details>

Note that (almost) all settings are css values, and, for example, you can use gradients/images/rgb,hsv, etc. in any combination as the background color for the minimized sidebar
This also means that instead of values in pixels you can use vw, vh, em, %, and any other css units.
I plan to extend this to light/dark theme customization

and yes, css injections work, but I think I'll leave it as it is, I just need to add validation

PetrK39 committed:
Basic theme manager
reimolemented auto dark mode feature
gelbooru darkmode button now force switches mode for session

PetrK39 committed:
css variables applying on DOM ready

PetrK39 committed:
gelbooru dark mode button tooltip
RMB clears force mode

PetrK39 committed:
rmb on gelbooru dark mode button clears force mode
it also have tooltip now

PetrK39 committed:
moved css background-color variable into themeManager

PetrK39 committed:
auto dark mode switch

PetrK39 committed:
configManager events test
separate files test

PetrK39 committed:
separate files

PetrK39 committed:
typo

PetrK39 committed:
made СonfigManager more detached

PetrK39 committed:
simplified preference keys

PetrK39 committed:
commented out WIP blacklistManager

PetrK39 committed:
commented out WIP blacklistManager

PetrK39 committed:
commented out WIP blacklistManager...

PetrK39 committed:
entire script now runs on DOM ready

PetrK39 committed:
working on the blacklist

PetrK39 committed:
PostItem now has rating and score
blacklist nwo works with infinite scrolling but it's too slow

PetrK39 committed:
css file rename

PetrK39 committed:
infinite scrolling is now separate module
but it still works slow
blacklist still buggy and VERY slow

PetrK39 committed:
blacklistManager now able to hide images
blackout, blur and show on hover features

PetrK39 commented:
Note to self: in incognito mode it actually pretty fast
check cache clearing, item search speed

Enchoseon commented:
Speedup might be from indexdb getting stored in memory rather than on-disk in your browser's incognito mode.

PetrK39 commented:
Also, I can't figure out history.pushState
It doesn't work until I open any post manually.  however, even so, behaviour of the history is far from expected

Enchoseon commented:
I'm unfamiliar with history.pushState and modifying browser history, but apparently the `"unused"` parameter that's being fed [`htmlDocument.title`](https://github.com/Enchoseon/gelbooru-overhaul-userscript/blob/a735e4cf51a59378af2dda31cbf018689ea29ea5/gelbooru-overhaul.infiniteScrolling.js#L107) is [nonstandard and deprecated](https://developer.mozilla.org/en-US/docs/Web/API/History/pushState#browser_compatibility). Not sure what the intended behavior is.

Also, regarding the navbar crash on the issue tracker and homepage, should we support the different navbar format at the expense of the CSS being different? _(see pending review for a screenshot)_ Or do we just want to skip inserting the configuration button entirely and not bother?

PetrK39 commented:
Yes, there are several situations where the script breaks the page.
Unfortunately, I don't have enough time right now to sit down and put everything back in order.

PetrK39 committed:
fixed crash on post history pages

PetrK39 committed:
blacklist's title now shows counts

PetrK39 committed:
blacklist now works fast
but there is still tiny delay before it applies filters
(post cache just was not working)

PetrK39 commented:
About browser history. I can confirm that it is a bug in Google Chrome as I have now checked it on Mozzila Firefox and everything works exactly as expected

PetrK39 committed:
collaps sidebar hiding its contents when not hover
not it can have any width

PetrK39 committed:
blacklist sidebar update cleanup

PetrK39 committed:
blacklist entries toggle

PetrK39 committed:
fixed utils.getThumbnails for posts with comments

PetrK39 committed:
blacklist editing
fixed AND and '_'
fixed check AND entry hit
tweaks in entries display

PetrK39 commented:
the script is significantly struggle when working with a *massive* blacklistlist in addition to a large number of posts

PetrK39 committed:
optimised blacklist applying
optimised entry toggle
fixed stupid typo in parseEntry
changed blacklist sidebar title logics, now it shows actual counts
new QoL utils.getThumbPostId function

PetrK39 commented:
It looks like we can't optimize the blacklist any further. The `checkPost` function has a complexity of O(post count * blacklist entries count), which results in 16464 comparisons for 392 entries per page[^1]
Since not all tags in my general blacklist belong to gelbooru I need to add an ignore list for tag namespaces
The saddest part is that there is still a delay between loading posts and hiding them. This can probably be fixed by using the magic of lazy

[^1]: It takes up to 2 seconds to apply the whole blacklist to 10 pages, which to me is acceptable

PetrK39 committed:
Keep selected blacklist and disabled entries

PetrK39 committed:
fixed modal window styles

PetrK39 committed:
configurable blacklist hidden images tweaks

PetrK39 committed:
fix to previous
fix blacklist sidebar on config save

Enchoseon commented on gelbooru-overhaul.user.js:
```diff
@@ -9,485 +9,1122 @@
+
+            this.scheduleCheckForThemeSwitch();
+
+            let darkModeButton = Object.values(document.querySelectorAll("#myTopnav a, #navbar a")).filter(i => i.getAttribute("href").includes("javascript:;"))[0];
+            darkModeButton.onclick = undefined;
+            darkModeButton.setAttribute("title", "Click to force switch dark mode for current session\nRight click to clear force mode");
+            darkModeButton.addEventListener("click", e => this.switchForceSessionMode());
+            darkModeButton.addEventListener("contextmenu", e => this.clearForceSessionMode());
```
```suggestion
            let darkModeButton = document.querySelector("#myTopnav a[onclick*='darkModeToggle()'][href*='javascript:;']");
            if (darkModeButton !== null) {
                darkModeButton.onclick = undefined;
                darkModeButton.setAttribute("title", "Click to force switch dark mode for current session\nRight click to clear force mode");
                darkModeButton.addEventListener("click", e => this.switchForceSessionMode());
                darkModeButton.addEventListener("contextmenu", e => this.clearForceSessionMode());
            }
```
Simplify selector for darkModeButton and fix crash on home page and issue tracker that stops page from loading.

PetrK39 replied:
straightforward fix d953a747f847d9ea5cc6cdf865f326fb9fa5935b

PetrK39 replied:
7fc9185ef8b53cf9bdae553f0cd872e1e8a9ea4f inspired by the suggested selector

Enchoseon reviewed (commented):
Mfw I left my review unpublished for a month and now everything I wrote is obsolete.

PetrK39 committed:
skip some tweaks on `Unknown` page type

PetrK39 committed:
simplify darkmode button selection

PetrK39 committed:
fixed select config value in config window
fixed missing updatInputValues for select config

PetrK39 committed:
made infinite scroll and blacklist lazy

PetrK39 commented:
The blacklist now works fast enough for infinite scrolling
But the namespaces for the tags are lost
Trying to make mass api queries leads to a lot of net errors

PetrK39 commented:
Since they are only used when downloading an additional tag file by RMB, they will only be resolved there, in a stable slow way

PetrK39 committed:
fix '_' for new tags

PetrK39 committed:
fix applyTweakRemoveTitle title set to 'null'

PetrK39 committed:
add extra blacklist safety on page load

PetrK39 committed:
fix removing blackout for undefined pages

PetrK39 committed:
force clear post item cache with RMB on 'Overhaul' button

Enchoseon commented on gelbooru-overhaul.configManager.js:
```diff
@@ -0,0 +1,462 @@
+                    },
+                    maxCache: {
+                        value: 420,
+                        name: "Max cahed post items",
```
```suggestion
                        name: "Max cached post items",
```
Minor spelling mistake.

PetrK39 replied:
In fact, there are a lot of *Engrish*
When I finished, I was going to ask you to check and correct the visible text

PetrK39 committed:
Update gelbooru-overhaul.configManager.js

Co-authored-by: Encho <38841444+Enchoseon@users.noreply.github.com>

PetrK39 committed:
fetch() now wrapped with repeatable queue, this fixes most of crashes   i believe
I've cleaned up async/promise things on my way but i barely know what I'm doung rn

PetrK39 committed:
Merge branch 'refactoring' of https://github.com/PetrK39/gelbooru-overhaul-userscript into refactoring

PetrK39 committed:
blacklist manager done

PetrK39 committed:
debug thing left

Enchoseon commented on gelbooru-overhaul.user.js:
```diff
@@ -9,485 +9,1142 @@
+
+        blacklistManager.setupManager(value);
+    }
+    function applyVariableBlacklist(value) {
+        blacklistManager.orderEntriesByHitCount = value;
```
```suggestion
    function applyVariableBlacklist(value) {
        if (![utils.pageTypes.GALLERY, utils.pageTypes.POST].includes(currentPageType)) return;
        blacklistManager.orderEntriesByHitCount = value;
```
Fixes a crash before the first contentful paint on pages without `#go-advBlacklistEntries`.

PetrK39 committed:
Update gelbooru-overhaul.user.js

Co-authored-by: Encho <38841444+Enchoseon@users.noreply.github.com>

PetrK39 committed:
fixed blacklist applying for 'more like this'

PetrK39 committed:
forgot to remove

PetrK39 committed:
fix for broken blacklisted items count

PetrK39 commented:
I have an idea:
Add feature to \*star\* some tags and pin them on top of the autocompletion list
But this would require to copy and modify a **copyrighted** code
Do we need this feature in the main branch?
<sub>For myself, I'm mostly browse gelbooru for one tag</sub>

PetrK39 commented:
<details>
<summary>There is copyright in the autocomplete.js</summary>
Grabbed from Danbooru, modified minimally to work on Gelbooru 
	https://github.com/danbooru/danbooru
	
	Copyright (c) 2013~2021, Danbooru Project
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

The views and conclusions contained in the software and documentation are those
of the authors and should not be interpreted as representing official policies,
either expressed or implied, of the FreeBSD Project.
</summary>
</details>

I don't know much about software licensing, but it doesn't seem to be a big deal if we change it.

Enchoseon commented:
I think the pinning feature is totally fine to have in the main branch. I'm planning on adding Hydrus integration whenever I have free time, so I'm lenient about cutting things for being "bloat". The only line I'll draw is on importing an external library (e.g. jquery, underscore, etc.) 

For the Danbooru code, we should add something like this to the LICENSE file:
```
The externally maintained libraries used by GelbooruOverhaul are:

- Autocomplete.js, located at danbooru/danbooru/app/javascript/src/javascripts/autocomplete.js, is licensed as follows:
  """
    Copyright (c) 2013~2021, Danbooru Project
    All rights reserved.
    
    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are met:
    
    1. Redistributions of source code must retain the above copyright notice, this
       list of conditions and the following disclaimer.
    2. Redistributions in binary form must reproduce the above copyright notice,
       this list of conditions and the following disclaimer in the documentation
       and/or other materials provided with the distribution.
    
    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
    ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
    ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
    (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
    ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
    
    The views and conclusions contained in the software and documentation are those
    of the authors and should not be interpreted as representing official policies,
    either expressed or implied, of the FreeBSD Project.
  """
```

Speaking of licensing, I just realized that GelbooruOverhaul doesn't have a license at the moment and is still All Rights Reserved. What license do you think this project should have? Personally, I don't really care whether the license is copyleft or permissive, but you have contributed 99.9% of the code so I'm interested in what you think.

PetrK39 commented:
I don't care too. Even though there is only my code, I probably would never have started a project like this if I hadn't seen your project.
For integration with Hydrus I'm using Hydrus Companion. And since GelbooruOverhaul aims only at one site I don't see any reason to just copy Hudrus Companion features. But that doesn't mean we can't improve their interaction.
But obviously, I didn't manage to work out the architecture for the script on the fly. You can see how much of a problem this is.

PetrK39 committed:
theme manager is separated

PetrK39 committed:
test tweak for infinite scroll appending same page twice

PetrK39 committed:
this.

PetrK39 committed:
user.js now contains only coordination code

PetrK39 committed:
can't close config window by clicking 'Overhaul' button again (fixed)

PetrK39 committed:
context.

PetrK39 committed:
attempted to work around rate limit
but blacklist will hang for about 30 seconds if you rapidly infScroll anyways

Enchoseon commented:
Alright, I'm very sorry for being deathly inactive and stonewalling this fantastic PR with my muteness. I'm just going to straight-up merge the changes you have made--remaining issues with lazy loading and infinite scrolling--for the sake of merging them (and doing a little housekeeping) because it has been several months and either this will never be merged or the refactor will need a refactor.
